python/frequenz/microgrid_component_graph/__init__.pyi

```diff
@@ -0,0 +1,264 @@
+
+from collections import abc
+from typing import AbstractSet, Generic, Protocol, TypeVar
+
+class InvalidGraphError(Exception):
+    """Exception type that will be thrown if graph data is not valid."""
+
+class FormulaGenerationError(Exception):
+    """Encountered during formula generation from the component graph."""
+
+class ComponentIdProtocol(Protocol):
+    def __int__(self) -> int:
+        """Get the integer representation of the Component ID."""
+
+class ComponentProtocol(Protocol):
+    @property
+    def id(self) -> ComponentIdProtocol:
+        """The Component ID"""
+
+class ConnectionProtocol(Protocol):
+    @property
+    def source(self) -> ComponentIdProtocol:
+        """The ID of the component at the start of the connection."""
+
+    @property
+    def destination(self) -> ComponentIdProtocol:
+        """The ID of the component at the end of the connection."""
+
+class ComponentGraphConfig:
+    """Configuration for the component graph."""
+
+    def __init__(
+        self,
+        allow_component_validation_failures: bool = False,
+        allow_unconnected_components: bool = False,
+        allow_unspecified_inverters: bool = False,
+        disable_fallback_components: bool = False,
+    ) -> None:
+        """Initialize this instance.
+
+        Args:
+            allow_component_validation_failures: Whether to allow validation errors on
+                components.  When this is `True`, the graph will be built even if there
+                are validation errors on the components.
+            allow_unconnected_components: Whether to allow unconnected components in the
+                graph, that are not reachable from the root.
+            allow_unspecified_inverters: Whether to allow untyped inverters in the
+                graph.  When this is `True`, inverters that have
+                `InverterType::Unspecified` will be assumed to be Battery inverters.
+            disable_fallback_components: Whether to disable fallback components in
+                generated formulas.  When this is `True`, the formulas will not include
+                fallback components.
+        """
+
+ComponentT = TypeVar("ComponentT", bound=ComponentProtocol)
+ConnectionT = TypeVar("ConnectionT", bound=ConnectionProtocol)
+
+class ComponentGraph(Generic[ComponentT, ConnectionT]):
+    """A graph representation of the electrical components in a microgrid."""
+
+    def __init__(
+        self,
+        components: AbstractSet[ComponentT],
+        connections: AbstractSet[ConnectionT],
+        config: ComponentGraphConfig = ComponentGraphConfig(),
+    ) -> None:
+        """Initialize this instance.
+
+        Args:
+            components: The list of components to build the graph with.
+            connections: The list of connections between the components.
+            config: The configuration for the component graph.
+        """
+
+    def component(self, component_id: ComponentIdProtocol) -> ComponentT:
```

Comment:
I would use `get_component()` instead, it always bothered me that the component graph in the SDK is not respecting the convention of functions being verbs. Or just use indexing (`__getitem__`, i.e. `component_graph[id]`), but maybe that's a bit weird.